### PR TITLE
upgrade: dynamically adjust batch size for automatic descriptor repair

### LINKED ISSUE #129949: upgrade: first upgrade to repair descriptor corruption can get stuck on txn retries
Currently the logic for repairing descriptors with corruption for upgrades work at a 1000 descriptors at a time, which can be problematic if a cluster has a large number of descriptors with corruption that are heavily contended. This can lead to this upgrade step hitting transaction retries forever, which can be observed with cluster queries like:
```
SELECT count(*) FROM (SELECT crdb_internal.repair_catalog_corruption(id, corruption) AS was_repaired FROM "".crdb_internal.kv_repairable_catalog_corruptions LIMIT 1000) WHERE was_repaired ```
```
We should automatically adjust this batch size, so that a reasonable number of descriptors is exercised when a txn retry is hit. This can be done at https://github.com/cockroachdb/cockroach/blob/cf787f691f3d757409f10bb2bb8cb1bec2b20562/pkg/upgrade/upgrades/first_upgrade.go#L168

Jira issue: CRDB-41792

gz#23238

Epic CRDB-40419

### PR DESCRIPTION
**Pull request by fqazi**
Previously, an upgrade was added to repair any descriptor corruptions detected. The defaults for this upgrade attempted to fix 1,000 descriptors within a transaction. This could end up failing with retry errors if the batch encountered any competing high-priority transactions like leasing. To address this, this patch dynamically adjusts the batch size either when excessive retries are observed or the batch of updates takes too long. Once the batch is small enough, only high-priority transactions are used.

Fixes: #129949

Release note (bug fix): Addressed a bug in the upgrade pre-condition for repairing descriptor corruption, which could lead to finalization being stuck

### CODE REVIEW

**Review by spilchen — COMMENTED** [review]
I like the overall approach and have no issue of it. I just had one nit and a clarification about the test.

> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-:-O5x0YgEBSY-zINTHFzv:b-bx9pp3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/upgrade/upgrades/first_upgrade.go` line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O5wtdncC5I1Wvpux-Jw:-O5wtdncC5I1Wvpux-Jx:bt8aois) ([raw file](https://github.com/cockroachdb/cockroach/blob/a244e6ca216c12777f53674d47da96286dd3ee7d/pkg/upgrade/upgrades/first_upgrade.go#L204)):*
> ```Go
> 					errors.HasType(err, (*timeutil.TimeoutError)(nil)) {
> 					batchSize = max(batchSize/2, 1)
> 					continue
> ```

nit: do you think it would be worthwhile to log something if we have to retry?
___
*[`pkg/upgrade/upgrades/first_upgrade_test.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O5x-PDr2AJqX--CUTVL:-O5x-PDr2AJqX--CUTVM:b-d4ac7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a244e6ca216c12777f53674d47da96286dd3ee7d/pkg/upgrade/upgrades/first_upgrade_test.go#L324)):*
> ```Go
> 	const qUpgrade = "SET CLUSTER SETTING version = crdb_internal.node_executable_version()"
> 	<-locksHeld
> 	tdb.Exec(t, qUpgrade)
> ```

If I understand correctly, the descriptor lease will expire after 30 seconds. The upgrade will wait for this expiration before it can complete successfully. Does this trigger a retry in the new code, or does it simply wait for the lease to expire synchronously? The query timeout for the repair is set to 1 minute. Should we consider extending the lease duration beyond 1 minute to ensure at least one retry occurs?

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/upgrade/upgrades/first_upgrade_test.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O5x-PDr2AJqX--CUTVL:-O5y2BPG33ZUzlrc46jH:bjdwacz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a244e6ca216c12777f53674d47da96286dd3ee7d/pkg/upgrade/upgrades/first_upgrade_test.go#L324)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

If I understand correctly, the descriptor lease will expire after 30 seconds. The upgrade will wait for this expiration before it can complete successfully. Does this trigger a retry in the new code, or does it simply wait for the lease to expire synchronously? The query timeout for the repair is set to 1 minute. Should we consider extending the lease duration beyond 1 minute to ensure at least one retry occurs?
</blockquote></details>

This is because the txn that was added below leases the system database descriptor, which the upgrade will need to modify for the version bump.  So, instead of making the WaitForOneVersion wait 5 miniutes for that lease to expire we only wait 30 seconds.

**Comment by annrpom** [comment]
👋 qq: I saw another batch size of 1000 in FirstUpgradeFromRelease -- would we want to do something similar there as here?

**Review by fqazi — COMMENTED** [review]
@annrpom Good point, yes let me extend the logic into that

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-:-O5yylA6Bk63jPylC276:b-z6wq4l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

**Review by spilchen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/upgrade/upgrades/first_upgrade.go` line 204 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O60xZZb54lz5CbZh6yp:-O60xZZb54lz5CbZh6yq:b-52x8x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/935aa6949b7735f2f233094613dc29c38372efa7/pkg/upgrade/upgrades/first_upgrade.go#L204)):*
> ```Go
> 					errors.HasType(err, (*timeutil.TimeoutError)(nil)) {
> 					batchSize = max(batchSize/2, 1)
> 					log.Infof(ctx, "reducing batch size of invalid_object repair query to %d (hipri=%T)",
> ```

nit: should it be lowercase t?



_Suggestion:_
```Go
log.Infof(ctx, "reducing batch size of invalid_object repair query to %d (hipri=%t)",
```
___
*[`pkg/upgrade/upgrades/first_upgrade_test.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O5x-PDr2AJqX--CUTVL:-O60zSN61-VmQytVuntg:b-wsvas3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a244e6ca216c12777f53674d47da96286dd3ee7d/pkg/upgrade/upgrades/first_upgrade_test.go#L324)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This is because the txn that was added below leases the system database descriptor, which the upgrade will need to modify for the version bump.  So, instead of making the WaitForOneVersion wait 5 miniutes for that lease to expire we only wait 30 seconds.
</blockquote></details>

Thanks for the explanation.

**Comment by fqazi** [comment]
RFAL @annrpom @spilchen

**Review by annrpom — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/upgrade/upgrades/first_upgrade.go` line 204 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O60xZZb54lz5CbZh6yp:-O6MxRLX6BKL9sVrtmUX:bqzpdiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/935aa6949b7735f2f233094613dc29c38372efa7/pkg/upgrade/upgrades/first_upgrade.go#L204)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: should it be lowercase t?
</blockquote></details>

^ in case you missed it; right now, this will print `... (hipri=Bool)`
___
*[`pkg/upgrade/upgrades/first_upgrade.go` line 118 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O6MxdxlBnK1SI1RMuKF:-O6MxdxlBnK1SI1RMuKG:bjrz7s9) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbbaacdf4bb575fbf425e6acf67a219b80eb2860/pkg/upgrade/upgrades/first_upgrade.go#L118)):*
> ```Go
> 				errors.HasType(err, (*timeutil.TimeoutError)(nil)) {
> 				batchSize = max(batchSize/2, 1)
> 				log.Infof(ctx, "reducing batch size of invalid_object repair query to %d (hipri=%T)",
> ```

nit: log.Infof(ctx, "reducing batch size of invalid_object repair query to %d (hipri=%t)" instead
___
*[`pkg/upgrade/upgrades/first_upgrade_test.go` line 430 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-O6MzGN9DKK_yQRi2vke:-O6MzGN9DKK_yQRi2vkf:b-24a0zb) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbbaacdf4bb575fbf425e6acf67a219b80eb2860/pkg/upgrade/upgrades/first_upgrade_test.go#L430)):*
> ```Go
> 	}
> 
> 	// Confirm the repairable descriptors exist. to avo
> ```

nit: cleanup comment

File: pkg/upgrade/upgrades/first_upgrade_test.go [hunk 1]
```diff
@@ -296,3 +356,87 @@ func TestFirstUpgradeRepair(t *testing.T) {
...
+			Server: &server.TestingKnobs{
+				DisableAutomaticVersionUpgrade: make(chan struct{}),
+				ClusterVersionOverride:         v0,
+			},
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+		},
+	})
+	defer testServer.Stopper().Stop(ctx)
+
+	const totalDescriptorsToTest = 5000
```
Comment (annrpom) [inline pkg/upgrade/upgrades/first_upgrade_test.go:395]: this is a lot of descriptors! nice!

are we concerned about adding this slow of a test to CI (idk how slow this one is compared to others)? would this make sense as a roachtest?

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @spilchen)

___
*[`pkg/upgrade/upgrades/first_upgrade_test.go` line 395 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130064#gh-1750729602:-O6N0Wrs6korntqXGGzz:bfr5emw) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbbaacdf4bb575fbf425e6acf67a219b80eb2860/pkg/upgrade/upgrades/first_upgrade_test.go#L395)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

this is a lot of descriptors! nice!

are we concerned about adding this slow of a test to CI (idk how slow this one is compared to others)? would this make sense as a roachtest?
</blockquote></details>

This is on the slower side and takes about a minute for this validation < 2 minutes. But I think that should be fine, WDYT?

**Review by annrpom — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/upgrade/upgrades/first_upgrade_test.go` line 395 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130064#gh-1750729602:-O6N2h_GEq3UJvYvlfBN:b-dn0pu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbbaacdf4bb575fbf425e6acf67a219b80eb2860/pkg/upgrade/upgrades/first_upgrade_test.go#L395)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This is on the slower side and takes about a minute for this validation < 2 minutes. But I think that should be fine, WDYT?
</blockquote></details>

Ok I agree -- I was thinking it was a bigger number

This LGTM! Nice work

**Review by spilchen — APPROVED** [review]
:lgtm: I like the changes you made in the second round.

> Reviewed 1 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130064#-:-O6N71nb5SZTwuIfZX3P:b16wljx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

**Comment by fqazi** [comment]
@annrpom @spilchen TFTR!

bors r+

**Pull request merged**